File: macos/Sources/BillingClient.swift
import Foundation

/// Grok Build–style credits fetch via CLI proxy JSON (Bearer OAuth).
enum BillingClient {
    static func fetch() async throws -> UsageResponse {
        let access = try await OidcSession.shared.ensureAccessToken()
        let tokens = OidcSession.shared.load()

        var req = URLRequest(url: OidcSession.billingURL)
        req.httpMethod = "GET"
        req.timeoutInterval = 15
        req.setValue("Bearer \(access)", forHTTPHeaderField: "Authorization")
        req.setValue(OidcSession.tokenHeader, forHTTPHeaderField: "X-XAI-Token-Auth")
        if let uid = tokens?.userId, !uid.isEmpty {
            req.setValue(uid, forHTTPHeaderField: "x-userid")
        }
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue("grld-macos-oauth", forHTTPHeaderField: "x-grok-client-version")
        req.setValue("GRLD-macOS/oauth", forHTTPHeaderField: "User-Agent")

        do {
            let (data, response) = try await URLSession.shared.data(for: req)
            if let http = response as? HTTPURLResponse {
                if http.statusCode == 401 || http.statusCode == 403 {
                    throw GrokDirectClient.ClientError.http(
                        http.statusCode,
                        "Session rejected — Sign In to Grok again"
                    )
                }
                if !(200...299).contains(http.statusCode) {
                    let body = String(data: data, encoding: .utf8) ?? ""
                    throw GrokDirectClient.ClientError.http(http.statusCode, body)
                }
            }
            return try parse(data)
        } catch let e as GrokDirectClient.ClientError {
            throw e
        } catch {
            throw GrokDirectClient.ClientError.network(error)
        }
    }

    static func parse(_ data: Data) throws -> UsageResponse {
        guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw GrokDirectClient.ClientError.parseFailed
        }
        guard let cfg = root["config"] as? [String: Any] else {
            throw GrokDirectClient.ClientError.parseFailed
        }

        let usedF: Double? = {
            if let n = cfg["creditUsagePercent"] as? NSNumber { return n.doubleValue }
            if let d = cfg["creditUsagePercent"] as? Double { return d }
            return nil
        }()
        let used = usedF.map { min(100, max(0, Int($0.rounded()))) }

        var periodType = "weekly"
        var periodStart: String?
        var periodEnd: String?
        if let period = cfg["currentPeriod"] as? [String: Any] {
            let t = period["type"] as? String ?? ""
            if t.uppercased().contains("MONTHLY") { periodType = "monthly" }
            periodStart = period["start"] as? String
            periodEnd = period["end"] as? String
        }
        if periodStart == nil { periodStart = cfg["billingPeriodStart"] as? String }
        if periodEnd == nil { periodEnd = cfg["billingPeriodEnd"] as? String }

        var products: [ProductUsage] = []
        if let arr = cfg["productUsage"] as? [[String: Any]] {
            for p in arr {
                let nameRaw = p["product"] as? String ?? "Unknown"
                // API often lists every product with no usagePercent (or 0).
                // Only keep categories that actually have used %.
                guard let n = p["usagePercent"] as? NSNumber else { continue }
                let pct = min(100, max(0, Int(n.doubleValue.rounded())))
                guard pct > 0 else { continue }
                let id = productId(for: nameRaw)
                let display = displayName(for: id, fallback: nameRaw)
                products.append(ProductUsage(product: id, name: display, usagePercent: pct))
            }
        }

        let hasSignal = used != nil
            || products.contains(where: { $0.usagePercent > 0 })
            || periodStart != nil
            || periodEnd != nil
        if !hasSignal {
            throw GrokDirectClient.ClientError.freePlan
        }

        let usedFinal = used ?? products.map(\.usagePercent).max() ?? 0
        let tier = (root["subscriptionTier"] as? String)
            ?? (root["subscription_tier"] as? String)

        let fetched = ISO8601DateFormatter().string(from: Date())
        return UsageResponse(
            ok: true,
            remainingPercent: max(0, min(100, 100 - usedFinal)),
            usedPercent: usedFinal,
            weeklyUsageAvailable: true,
            tierName: tier,
            currentPeriod: UsagePeriod(type: periodType, start: periodStart, end: periodEnd),
            productUsage: products.sorted { $0.usagePercent > $1.usagePercent },
            fetchedAt: fetched,
            cached: false
        )
    }

    private static func productId(for raw: String) -> Int {
        let s = raw.lowercased()
            .replacingOccurrences(of: "_", with: "")
            .replacingOccurrences(of: "-", with: "")
            .replacingOccurrences(of: " ", with: "")
        if s.contains("build") { return 2 }
        if s.contains("chat") { return 4 }
        if s.contains("imagine") || s.contains("image") { return 5 }
        if s.contains("plugin") { return 3 }
        if s.contains("voice") { return 6 }
        if s.contains("api") { return 1 }
        if s.contains("third") || s.contains("3rd") { return 0 }
        return 100 + abs(s.hashValue % 200)
    }

    private static func displayName(for id: Int, fallback: String) -> String {
        switch id {
        case 0: return "3rd Party"
        case 1: return "API"
        case 2: return "Grok Build"
        case 3: return "Grok Plugins"
        case 4: return "Chat"
        case 5: return "Imagine"
        case 6: return "Voice"
        default:
            // "GrokBuild" → "Grok Build"
            var out = ""
            for (i, ch) in fallback.enumerated() {
                if i > 0, ch.isUppercase, fallback[fallback.index(fallback.startIndex, offsetBy: i - 1)].isLowercase {
                    out.append(" ")
                }
                out.append(ch)
            }
            return out.isEmpty ? fallback : out
        }
    }
}

File: android/app/src/main/java/com/blankspeaker/grld/BillingClient.kt
package com.blankspeaker.grld

import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import kotlin.math.roundToInt

/**
 * Grok Build–style credits fetch:
 * GET https://cli-chat-proxy.grok.com/v1/billing?format=credits
 * Authorization: Bearer <oauth access token>
 */
object BillingClient {

    private const val TAG = "GRLD-Billing"

    fun fetch(context: Context): UsageResponse {
        val access = OidcSession.ensureAccessToken(context)
            ?: throw GrokClient.ClientError.NotSignedIn()
        val tokens = OidcSession.load(context)
            ?: throw GrokClient.ClientError.NotSignedIn()

        val conn = (URL(OidcSession.BILLING_URL).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 12_000
            readTimeout = 15_000
            setRequestProperty("Authorization", "Bearer $access")
            setRequestProperty("X-XAI-Token-Auth", OidcSession.TOKEN_HEADER)
            if (tokens.userId.isNotBlank()) {
                setRequestProperty("x-userid", tokens.userId)
            }
            setRequestProperty("Accept", "application/json")
            setRequestProperty("x-grok-client-version", "grld-android-oauth-test")
            setRequestProperty("User-Agent", "GRLD-Android/oauth-test")
        }
        try {
            val code = conn.responseCode
            val raw = (if (code in 200..299) conn.inputStream else conn.errorStream)
                ?.readBytes()?.toString(StandardCharsets.UTF_8).orEmpty()
            if (code == 401 || code == 403) {
                throw GrokClient.ClientError.Http(code, "Session rejected — Sign In again")
            }
            if (code !in 200..299) {
                throw GrokClient.ClientError.Http(code, raw.take(200))
            }
            return parse(raw)
        } catch (e: GrokClient.ClientError) {
            throw e
        } catch (e: Exception) {
            Log.w(TAG, "billing fetch failed", e)
            throw GrokClient.ClientError.Network(e)
        } finally {
            conn.disconnect()
        }
    }

    fun parse(jsonText: String): UsageResponse {
        val root = JSONObject(jsonText)
        val cfg = root.optJSONObject("config")
            ?: throw GrokClient.ClientError.ParseFailed()

        val usedF = when {
            cfg.has("creditUsagePercent") && !cfg.isNull("creditUsagePercent") ->
                cfg.getDouble("creditUsagePercent")
            else -> null
        }
        val used = usedF?.roundToInt()?.coerceIn(0, 100)

        val periodObj = cfg.optJSONObject("currentPeriod")
        val period = if (periodObj != null) {
            val type = periodObj.optString("type", "USAGE_PERIOD_TYPE_WEEKLY")
            val kind = when {
                type.contains("MONTHLY", ignoreCase = true) -> "monthly"
                else -> "weekly"
            }
            UsagePeriod(
                type = kind,
                start = periodObj.optString("start", null).takeIf { !it.isNullOrBlank() }
                    ?: cfg.optString("billingPeriodStart", null).takeIf { !it.isNullOrBlank() },
                end = periodObj.optString("end", null).takeIf { !it.isNullOrBlank() }
                    ?: cfg.optString("billingPeriodEnd", null).takeIf { !it.isNullOrBlank() }
            )
        } else {
            UsagePeriod(
                type = "weekly",
                start = cfg.optString("billingPeriodStart", null).takeIf { !it.isNullOrBlank() },
                end = cfg.optString("billingPeriodEnd", null).takeIf { !it.isNullOrBlank() }
            )
        }

        val products = mutableListOf<ProductUsage>()
        val arr = cfg.optJSONArray("productUsage")
        if (arr != null) {
            for (i in 0 until arr.length()) {
                val p = arr.optJSONObject(i) ?: continue
                val nameRaw = p.optString("product", "").ifBlank { "Unknown" }
                // API often lists every product with no usagePercent (or 0).
                // Only keep categories that actually have used %.
                if (!p.has("usagePercent") || p.isNull("usagePercent")) continue
                val pct = p.getDouble("usagePercent").roundToInt().coerceIn(0, 100)
                if (pct <= 0) continue
                val id = productIdForName(nameRaw)
                products.add(
                    ProductUsage(
                        product = id,
                        name = ProductColors.displayName(id).let { dn ->
                            if (dn.startsWith("Other")) humanName(nameRaw) else dn
                        },
                        usagePercent = pct
                    )
                )
            }
        }

        // Free plan / no weekly pool: no percent and no products with usage
        val hasSignal = used != null || products.any { it.usagePercent > 0 } ||
            period.start != null || period.end != null
        if (!hasSignal) throw GrokClient.ClientError.FreePlan()

        val usedFinal = used ?: products.maxOfOrNull { it.usagePercent } ?: 0
        val tier = root.optString("subscriptionTier", null)
            ?.takeIf { it.isNotBlank() }
            ?: root.optString("subscription_tier", null)?.takeIf { it.isNotBlank() }

        return UsageResponse(
            ok = true,
            remainingPercent = (100 - usedFinal).coerceIn(0, 100),
            usedPercent = usedFinal.coerceIn(0, 100),
            weeklyUsageAvailable = true,
            tierName = tier,
            currentPeriod = period,
            productUsage = products.sortedByDescending { it.usagePercent },
            fetchedAt = isoNow(),
            cached = false
        )
    }

    private fun productIdForName(raw: String): Int {
        val s = raw.lowercase(Locale.US)
            .replace("_", "")
            .replace("-", "")
            .replace(" ", "")
        // Exact known API names first (avoids substring traps like "appbuilder" ⊃ "build").
        return when (s) {
            "grokbuild", "build" -> ProductColors.BUILD.id
            "grokappbuilder", "appbuilder" -> ProductColors.APP_BUILDER.id
            "grokchat", "chat" -> ProductColors.CHAT.id
            "grokimagine", "imagine", "image", "images" -> ProductColors.IMAGINE.id
            "grokplugins", "plugins", "plugin" -> ProductColors.PLUGINS.id
            "voice", "grokvoice" -> ProductColors.VOICE.id
            "api", "grokapi" -> ProductColors.API.id
            "thirdparty", "3rdparty", "third" -> ProductColors.THIRD_PARTY.id
            else -> when {
                s.contains("appbuilder") -> ProductColors.APP_BUILDER.id
                s.contains("third") || s.contains("3rd") -> ProductColors.THIRD_PARTY.id
                s.contains("plugin") -> ProductColors.PLUGINS.id
                s.contains("imagine") || s.contains("image") -> ProductColors.IMAGINE.id
                s.contains("voice") -> ProductColors.VOICE.id
                s.contains("chat") -> ProductColors.CHAT.id
                s.contains("api") -> ProductColors.API.id
                // "build" but not "builder" (GrokAppBuilder must not map to Grok Build).
                s.contains("build") && !s.contains("builder") -> ProductColors.BUILD.id
                else -> {
                    // Stable pseudo-id for unknowns so palette still works
                    100 + (s.hashCode() and 0xFF)
                }
            }
        }
    }

    private fun humanName(raw: String): String {
        return raw
            .replace(Regex("([a-z])([A-Z])"), "$1 $2")
            .replace('_', ' ')
            .trim()
            .ifBlank { raw }
    }

    private fun isoNow(): String {
        val df = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
        df.timeZone = TimeZone.getTimeZone("UTC")
        return df.format(Date())
    }
}
